### PR TITLE
WIP: Create triangle class

### PR DESCRIPTION
**Pull request by roark-z**
### Description

Added triangle class under geom directory

### Testing Done

Added unit tests for functions in triangle class

### Resolved Issues

resolves `#417`

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [ ] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/geom/triangle.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        // one-definition rule, use anonymous namespace for
+        // intersects() in geom/util.h
+        if (::intersects(ray, seg))
+        {
+            num_intersections++;
+        }
+    }
+
+    // if the ray intersects the triangle exactly once
+    // it is inside the polygon, otherwise it is outside the triangle
```
Comment (EvanMorcom) [inline src/thunderbots/software/geom/triangle.cpp]: What if the intersection is right where 2 segments meet?

File: src/thunderbots/software/geom/triangle.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
+#include "geom/triangle.h"
+
+#include <unordered_set>
+
+#include "geom/util.h"
+
+Triangle::Triangle(Point p0, Point p1, Point p2)
```
Comment (EvanMorcom) [inline src/thunderbots/software/geom/triangle.cpp]: Maybe unnecessary, but what if the points specified do not form a triangle? Like if 2 points are the same

**Review by EvanMorcom — DISMISSED** [review]
Lookin' good. Left a couple comments :+1:

File: src/thunderbots/software/geom/triangle.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/geom/triangle.cpp]: I suppose you'd just end up with a Triangle with 0 area. Not sure if there's much we can do to prevent it, and I don't think it's necessary.

File: src/thunderbots/software/geom/triangle.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/geom/triangle.cpp]: I'm not sure what would happen. This case definitely needs a few unit tests.

File: src/thunderbots/software/geom/triangle.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+     * Returns the points that form the triangle.
+     * @return the points that form the triangle.
+     */
+    const std::vector<Point>& getPoints() const;
+
+   private:
+    // the line segments that form the polygon
+    std::vector<Segment> segments;
+    std::vector<Point> points;
+};
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/triangle.h]: nit: no blank line at the end of the file

**Review by MathewMacDougall — DISMISSED** [review]

**Review by MathewMacDougall — DISMISSED** [review]
I'd like to see some more tests for the `containsPoint` function to cover the case that Evan mentioned, where the ray we use to check how many sides of the triangle we intersect goes exactly across a vertex. If 2 segments are detected as intersected, we could end up with false positives and false negatives.

Eg. Cases where the point is outside the triangle and the ray intersects a vertex, and the same where the point is inside the triangle.

Otherwise the rest looks good.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
New tests look good, but looks like CI is failing because some of the KickAction tests are now not returning the expected Intent.

This is probably because the Triangle now being used for the "behind ball region" behaves slightly differently now checking if a point is inside (is my guess). 

That will need to be fixed but otherwise LGTM

**Comment by MathewMacDougall** [comment]
Whoops didn't mean to close this. Looks like the tests are still failing. Make sure to check why the tests are failing with the old cases (the debugger is great for this). I'm not sure if it's exposing edge cases in the new code, or if we're just not using the new code correctly now (since you changed the implementation to use Triangle now).

File: src/thunderbots/software/test/ai/hl/stp/action/kick_action.cpp [hunk 4]
```diff
@@ -19,18 +19,18 @@ TEST(KickActionTest, robot_behind_ball_kicking_towards_positive_x_positive_y)
...
-        EXPECT_EQ(0, kick_intent.getRobotId());
-        EXPECT_EQ(Point(0, 0), kick_intent.getKickOrigin());
-        EXPECT_EQ(Angle::zero(), kick_intent.getKickDirection());
-        EXPECT_EQ(5.0, kick_intent.getKickSpeed());
-    }
-    catch (...)
-    {
-        ADD_FAILURE() << "Kick intent not returned by Kick Action!";
-    }
+    //    try
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/ai/hl/stp/action/kick_action.cpp]: The try-catch should be uncommented so we get more descriptive test failures here

File: src/thunderbots/software/ai/hl/stp/action/kick_action.cpp [hunk 5]
```diff
@@ -90,10 +91,11 @@ void KickAction::calculateNextIntent(IntentCoroutine::push_type &yield)
             behind_ball_vertex_A + behind_ball.norm(size_of_region_behind_ball) -
             behind_ball.perp().norm(size_of_region_behind_ball / 2);
 
-        Polygon behind_ball_region =
-            Polygon({behind_ball_vertex_A, behind_ball_vertex_B, behind_ball_vertex_C});
-
-        bool robot_behind_ball = behind_ball_region.containsPoint(robot->position());
+        // Triangle behind_ball_region =
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/hl/stp/action/kick_action.cpp]: Should this commented-out code be being used?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Few small comments but otherwise looks good.

Seeing some odd CI failures
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuite tests="1" failures="1" time="1" errors="0" name="gtest-action_test.xml">
  <testcase name="test_ran" status="run" time="1" classname="Results">
    <failure message="Unable to find test results for gtest-action_test.xml, test did not run.
Expected results in /root/Software/build/test_results/thunderbots/gtest-action_test.xml" type=""/>
  </testcase>
</testsuite>
-------------------------------------------------
Full test results for 'build/test_results/thunderbots/MISSING-gtest-triangle_test.xml'
-------------------------------------------------
<?xml version="1.0" encoding="UTF-8"?>
<testsuite tests="1" failures="1" time="1" errors="0" name="gtest-triangle_test.xml">
  <testcase name="test_ran" status="run" time="1" classname="Results">
    <failure message="Unable to find test results for gtest-triangle_test.xml, test did not run.
Expected results in /root/Software/build/test_results/thunderbots/gtest-triangle_test.xml" type=""/>
  </testcase>
</testsuite>
```

I've restarted CI to see if these issues were just a one-off issue.

File: src/thunderbots/software/geom/triangle.cpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#include "geom/triangle.h"
+
+#include <algorithm>
+#include <unordered_set>
+
+#include "geom/util.h"
+
+Triangle::Triangle(Point p0, Point p1, Point p2) : Polygon(std::vector<Point>{p0, p1, p2})
+{
+    segments[0] = Segment{p0, p1};
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/triangle.cpp:10]: You can't assign elements of the vector without allocating / initializing them first. This is why the test are failing, because this causes a segfault. The memory this segment is being assigned to hasn't been allocated yet.

This should probably use `emplace_back` instead to append the value to the vector. This goes for the other functions in this file too that are setting up `segments`

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Figured out why the tests are failing. See comment below.

**Comment by MathewMacDougall** [comment]
@SupoWupo any updates here?

**Comment by MathewMacDougall** [comment]
@SupoWupo 	you'll need to pull from `upstream/master` to fix the conflicts, but that should also fix any CI issues, and CI builds should be faster now.

**Comment by MathewMacDougall** [comment]
@SupoWupo I'm going to close this PR if we're no longer making progress. We can revisit these changes next year. Please let me know if you're still working on this or not.
